client/src/main/java/io/split/client/impressions/ImpressionObserver.java

```diff
@@ -0,0 +1,38 @@
+package io.split.client.impressions;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.split.client.dtos.KeyImpression;
+
+/*
+According to guava's docs (https://guava.dev/releases/18.0/api/docs/com/google/common/annotations/Beta.html),
+the @Beta decorator only means that the api is not frozen, and has nothing to do with behaviour stability, but
+rather to a non-frozen API which may introduce breaking changes at any time in future versions.
+Since the library is shaded and should not be exposed to users of the SDK, it's safe to use it here.
+ */
+
+@SuppressWarnings("UnstableApiUsage")
+public class ImpressionObserver {
+
+    private final Cache<Long, Long> _cache;
+
+    public ImpressionObserver(long size) {
+        _cache = CacheBuilder.newBuilder()
+                .maximumSize(size)
+                .build();
+    }
+
+    public Long testAndSet(KeyImpression impression) {
+        if (null == impression) {
+            return null;
+        }
+
+        Long hash = ImpressionHasher.process(impression);
+
+        synchronized(this) {
```

Comment:
this can't be here. It'll be too slow. Better to use the compute method from guava cache.

Reply:
I guess the caller is single thread correct?

Reply:
I'm not familiar with `compute` method, i'll look into it.
It's single threaded, but I added the synchronized block, in order to avoid surprises in the future, are those statements really that slow? the hashing computation is done outside of the critical region